xwiki-platform-core/xwiki-platform-refactoring/xwiki-platform-refactoring-default/src/main/java/org/xwiki/refactoring/internal/ResourceReferenceRenamer.java

```diff
@@ -155,24 +195,48 @@ private boolean updateAbsoluteResourceReference(ResourceReference resourceRefere
     }
 
     private <T extends EntityReference> boolean updateRelativeResourceReference(ResourceReference resourceReference,
-        T oldReference, T newReference)
+        T oldReference, T newReference, Map<EntityReference, EntityReference> updatedEntities)
     {
         boolean result = false;
 
         EntityReference linkEntityReference = this.entityReferenceResolver.resolve(resourceReference, null,
             newReference);
-
-        if (newReference.equals(linkEntityReference.extractReference(EntityType.DOCUMENT))) {
-            // If the link is relative to the containing document we don't modify it
-            return false;
-        }
-
         // current link, use the old document's reference to fill in blanks.
         EntityReference oldLinkReference =
             this.entityReferenceResolver.resolve(resourceReference, null, oldReference);
+        EntityReference absoluteResolvedReference = this.entityReferenceResolver.resolve(resourceReference, null);
 
-        // If the new and old link references don`t match, then we must update the relative link.
-        if (!linkEntityReference.equals(oldLinkReference)) {
+        boolean docExists = false;
+        EntityType entityType = linkEntityReference.getType();
+        DocumentReference documentReference = null;
+        DocumentReference documentReferenceTarget = null;
+        if (entityType == EntityType.DOCUMENT || entityType.getAllowedParents().contains(EntityType.DOCUMENT)) {
+            documentReference =
+                (DocumentReference) oldLinkReference.extractReference(EntityType.DOCUMENT);
+            documentReferenceTarget =
+                (DocumentReference) linkEntityReference.extractReference(EntityType.DOCUMENT);
+            docExists = checkIfDocExists(documentReference);
+        } else {
+            docExists = true;
+        }
+
+        boolean anyMatchInMovedReferences =
+            (oldLinkReference.hasParent(oldReference)
+                || (documentReferenceTarget != null
+                && documentReferenceTarget.equals(updatedEntities.get(documentReference))));
+
+        // We should update the reference if:
+        //  - it's relative: the resolution of the reference without any parameter, and the resolution of the reference
+        //                   with the new reference should give different results
```

Comment:
Why do you need this check? Isn't the explanation the same as what is already checked by the third condition?

Reply:
No the conditions are very different. This one is only checking that the current link is relative: i.e. the condition will never be met if I have a link to `xwiki:MySpace.MyPage`, but it will be met if I have `MySpace.MyPage`. 
The third condition is about checking that the link points to a different target because of the rename, so if I have a link `MySpace.MyPage` in page `xwiki:Bla.MySpace` and I renamed `Bla` to `Foo` the resolution was `xwiki:Bla.MySpace.MyPage` and it's now `xwiki:Foo.MySpace.MyPage`: in such case the condition is met.

Reply:
> No the conditions are very different. This one is only checking that the current link is relative: i.e. the condition will never be met if I have a link to `xwiki:MySpace.MyPage`, but it will be met if I have `MySpace.MyPage`.

That's true, but what's relevant here is that the resolution relative to the old location is different and that's already checked by the third condition. So unless there is a different explanation why this check is needed, I think it could also just be removed. A possible reason could be that the absolute rename step already adapted the link, and we want to prevent it from being adapted again?

Reply:
> A possible reason could be that the absolute rename step already adapted the link, and we want to prevent it from being adapted again?

It's possible I needed it for that, to be honest I don't recall exactly. 

I'll try to run again my test without that specific condition just to see if one of them is failing.

Reply:
> I'll try to run again my test without that specific condition just to see if one of them is failing.

Ok apparently I can indeed discard that condition safely, at least for my tests to pass.